### ISSUE TITLE
BSOD when trying to set up a VXLAN tunnel

### ISSUE THREAD

**Issue by aserdean** [openvswitch/ovs-issues#20]
This issue happens by trying to set up a VXLAN tunnel between two Hyper-V nodes. The crash is in the logging part and only affects the Debug build. In Release mode a different crash happens reported separately.

Steps to reproduce the issue on two Hyper-V nodes (H1 and H2):

On H1:
ovs-vsctl --db=tcp:127.0.0.1:8888 add-br br0
ovs-vsctl --db=tcp:127.0.0.1:8888 add-port br0 external.1
ovs-vsctl --db=tcp:127.0.0.1:8888 add-port br0 internal
ovs-vsctl --db=tcp:127.0.0.1:8888 add-port br0 vxlan
ovs-vsctl --db=tcp:127.0.0.1:8888 set Inteface vxlan type=vxlan
ovs-vsctl --db=tcp:127.0.0.1:8888 set Inteface vxlan options:remote_ip=10.13.8.160
ovs-vsctl --db=tcp:127.0.0.1:8888 set Inteface vxlan options:local_ip=10.13.8.101
ovs-vsctl --db=tcp:127.0.0.1:8888 set Inteface vxlan options:in_key=flow
ovs-vsctl --db=tcp:127.0.0.1:8888 set Inteface vxlan options:out_key=flow

On H2:
ovs-vsctl --db=tcp:127.0.0.1:8888 add-br br0
ovs-vsctl --db=tcp:127.0.0.1:8888 add-port br0 external.1
ovs-vsctl --db=tcp:127.0.0.1:8888 add-port br0 internal
ovs-vsctl --db=tcp:127.0.0.1:8888 add-port br0 vxlan
ovs-vsctl --db=tcp:127.0.0.1:8888 set Inteface vxlan type=vxlan
ovs-vsctl --db=tcp:127.0.0.1:8888 set Inteface vxlan options:remote_ip=10.13.8.101
ovs-vsctl --db=tcp:127.0.0.1:8888 set Inteface vxlan options:local_ip=10.13.8.160
ovs-vsctl --db=tcp:127.0.0.1:8888 set Inteface vxlan options:in_key=flow
ovs-vsctl --db=tcp:127.0.0.1:8888 set Inteface vxlan options:out_key=flow

Issue a ping command on H2 to H1:
ping -t 10.13.8.101

Here is the stack trace and Bugcheck Analysis.
http://pastebin.com/cUsh73Hf

Workaround provided by Nithin under:  https://github.com/openvswitch/ovs/issues/9

**Comment by nithinrajub** [comment]
I spent some time on this today.

As mentioned in openvswitch/ovs#9, the issue here is that there is a loop created by adding the tunnel port to the PIF bridge. When a packet sent out from the internal port (VTEP) reaches the tunnel port, the tunnel port promptly encapsulates the packet, and re-injects the packet onto the bridge with the source port as the VTEP port, and so on and so forth.

There are two flaws here:
1. There's an implicit assumption that there is another br-pif in the tunneling code. ie. it tries to do a flow lookup for a second time when it re-injects the packet. I have to see how true this holds when say there's no PIF bridge at all. This might be OK though.
2. The Tunneling code should not to encapsulation and re-inject the packet if the source port is the same as the VTEP port.

I looked in dpif-linux.c to check if there's a source port attributes passed down to the kernel, and voila!
2758 /\* Generate ODP flow key from the given packet metadata _/  
2759 void  
2760 odp_key_from_pkt_metadata(struct ofpbuf *buf, const struct pkt_metadata *md)  
2761 {  
[...]
2770     /_ Add an ingress port attribute if 'odp_in_port' is not the magical  
2771      \* value "ODPP_NONE". */  
2772     if (md->in_port.odp_port != ODPP_NONE) {  
2773         nl_msg_put_odp_port(buf, OVS_KEY_ATTR_IN_PORT, md->in_port.odp_port); 
2774     }  
2775 }

In the kernel datapath, we need to check if the source port is a VIF port and only perform encapsulation if so. That should fix the issue. I can send out a patch for this on the existing datapath which interfaces with the legacy dpif-windows.c or wait till the new netlink based dpif is implemented. Either way, this can be fixed easily.
